tests/Unit/DataStructures/Test_VectorImpl.cpp

```diff
@@ -0,0 +1,25 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include <array>
+#include <complex>
+
+#include "DataStructures/DataVector.hpp"  // IWYU pragma: keep
+#include "DataStructures/VectorImpl.hpp"  // IWYU pragma: keep
+#include "Utilities/TypeTraits.hpp"
+
+/// [vector_base_type_example]
+static_assert(cpp17::is_same_v<vector_base_type_t<DataVector>, double>,
+              "Failed testing type trait vector_base_type");
+
+static_assert(
+    cpp17::is_same_v<vector_base_type_t<std::array<DataVector, 2>>, double>,
+    "Failed testing type trait vector_base_type");
+
+static_assert(cpp17::is_same_v<vector_base_type_t<std::complex<double>*>,
+                               std::complex<double>>,
+              "Failed testing type trait vector_base_type");
+
+static_assert(cpp17::is_same_v<vector_base_type_t<DataVector&>, double>,
```

Comment:
Seeing this makes me think we should not allow `VectorType` to be a pointer or a reference. I don't know what that really means and I think most of the implementation would just blow up anyway. I'm fine adding a `static_assert` in a followup PR.

Reply:
Regarding pointer, I'm fine with, but because we specifically test references in the 'reference wrapping' context for math tests, disallowing references would significantly impact the usefulness of this function for the later commit in this PR. Can you elaborate on your objection?

Reply:
What I mean is that I don't think `VectorImpl<double, DataVector&>` should be allowed because I don't know what that means (same with pointer). I'm not objecting to the trait handling those cases.

Reply:
oh, woa. Yeah, I agree that's going to lead to confusing behavior. :+1:

Reply:
Yep, but like I said, I think that should be deferred to a later PR :)